subprojects/messaging/src/main/java/org/gradle/internal/serialize/DefaultSerializerRegistry.java

```diff
@@ -28,8 +28,11 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.TreeMap;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 public class DefaultSerializerRegistry implements SerializerRegistry {
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
```

Comment:
I would not use a single lock to protect the two different data structures.
A lock seems the only way to properly guard the `TreeMap` indeed. But for the `HashSet`, I would use a `ConcurrentHashMap` with a a constant value and the `Class<?>` as the key.

Reply:
As commented by @tjni `jConcurrentHashMap#newKeySet()` would do the job indeed.

Reply:
Done!